### PR TITLE
fix: prioritize .taskmaster in parent directories over other project markers

### LINKED ISSUE #1349: Bug: findProjectRoot() prioritizes other project markers over parent .taskmaster directories
## Bug Description

When running task-master commands from subdirectories that contain other project markers (like `.git`, `go.mod`, `package.json`), `findProjectRoot()` stops at the first project marker found instead of continuing to search for a `.taskmaster` directory in parent directories.

This prevents the intended behavior for multi-repo monorepo setups where a single `.taskmaster` directory at the root should track work across multiple sub-repositories.

## Steps to Reproduce

1. Create a directory structure like:
   ```
   /project-root/
   ├── .taskmaster/          # Task Master tracking all work
   └── sub-repo/
       ├── .git/             # Git repository marker
       └── go.mod            # Go project marker
   ```

2. Initialize Task Master at the root:
   ```bash
   cd /project-root
   task-master init
   task-master add-task --prompt="Test task"
   ```

3. Navigate to the sub-repository and run a command:
   ```bash
   cd sub-repo
   task-master list
   ```

## Expected Behavior

Task Master should find the parent `.taskmaster` directory and display the tasks from the root project.

## Actual Behavior

Task Master creates a NEW `.taskmaster` directory in `sub-repo/` instead of finding the parent, resulting in:
```
Listing tasks from: /project-root/sub-repo/.taskmaster/tasks/tasks.json
⚠ No tasks found matching the criteria.
```

## Root Cause

The `findProjectRoot()` function in `src/utils/path-utils.js` currently:
1. Checks for ALL project markers (Task Master + generic) in the current directory
2. Returns immediately when ANY marker is found
3. Never checks parent directories for Task Master markers if a generic marker exists in current directory

Since `.taskmaster`, `.git`, `go.mod`, etc. are all in the same priority list, whichever is checked first wins, preventing parent traversal.

## Proposed Solution

PR #[number] implements a two-pass search strategy:
1. **First pass**: Search ALL parent directories for Task Master-specific markers only
2. **Second pass**: If no Task Master markers found, search for other project markers

This ensures `.taskmaster` in parent directories always takes precedence over other project markers in subdirectories.

## Impact

This affects:
- Multi-repo monorepo setups
- Microservices architectures with a central Task Master
- Any workflow where task-master commands are run from nested repositories
- Projects using git worktrees with a central `.taskmaster`

## Environment

- Task Master version: 0.30.0+
- Platform: All (macOS, Linux, Windows)
- Related to: PR #1302 (which added parent traversal but didn't address marker priority)

## Workaround

Currently, users must create symlinks in subdirectories:
```bash
cd sub-repo
ln -s ../.taskmaster .taskmaster
```

### PR DESCRIPTION
**Pull request by bjcoombs**
## Summary

Fixes #1349

Ensures `.taskmaster` directories in parent folders take priority over other project markers (`.git`, `go.mod`, etc.) in subdirectories.

**Problem**: Running task-master from subdirectories with their own project markers would create duplicate `.taskmaster` directories instead of finding the parent.

**Solution**: Two-pass search strategy - first pass searches ALL parents for Task Master markers only, second pass searches for generic project markers if none found.

## Implementation

Modified `findProjectRoot()` in `src/utils/path-utils.js`:

1. **First Pass**: Traverse parents looking ONLY for Task Master-specific markers (`.taskmaster/`, `.taskmaster/config.json`, `.taskmasterconfig`)
2. **Second Pass**: If no TM markers found, search for generic markers (`.git`, `package.json`, `go.mod`, etc.)
3. **Includes root directory**: Fixed bug where filesystem root was never checked
4. **No generic markers in first pass**: Moved `tasks.json` and `tasks/tasks.json` to second pass (not TM-specific)

## Key Changes

- Split project markers into two arrays (TM-specific vs generic)
- Implemented two-pass traversal with separate marker lists
- Enhanced test coverage with monorepo scenarios
- Updated coverage thresholds for utils (90%)

## Use Cases

- **Monorepos**: Single `.taskmaster` at root, multiple Git repos in subdirectories
- **Microservices**: Central task tracking across service repositories  
- **Git worktrees**: Worktree directories use parent `.taskmaster`

## Example

```bash
# Structure:
/project/.taskmaster/
/project/service-a/.git
/project/service-a/go.mod

# Before: Creates /project/service-a/.taskmaster/
# After:  Uses /project/.taskmaster/
cd /project/service-a && task-master list
```

## Backward Compatibility

✅ Fully compatible - only affects cases where parent `.taskmaster` exists AND subdirectory has other markers (desired behavior change)

## Testing

- ✅ All 15 unit tests passing
- ✅ New tests for monorepo scenarios
- ✅ Root directory edge case covered
- ✅ Generic task file handling verified

---

**Related**: Builds on #1302 (parent traversal) with proper marker prioritization

## Summary by CodeRabbit

* **New Features**
  * CLI now auto-detects project root for many commands; autopilot start adds optional maxAttempts.
  * Core exposes centralized path/marker constants and public project-root utilities.

* **Bug Fixes**
  * Project-root detection prefers parent .taskmaster markers and has safer traversal/fallback behavior.
  * Dev startup loads .env from the detected project root for consistent config resolution.

* **Tests**
  * Added comprehensive project-root tests and improved isolation; legacy unit suite removed.

* **Chores**
  * Added per-path coverage thresholds and deprecation notes for legacy wrappers.

---

> [!NOTE]
> Prioritizes parent `.taskmaster` when detecting project roots, exposes new core utils/constants, updates CLI commands to auto-detect roots, and adds targeted tests and config tweaks.
> 
> - **Core**:
>   - **Project Root Detection**: Add `findProjectRoot()` with two-pass strategy prioritizing `.taskmaster` over generic markers; add `normalizeProjectRoot()` in `packages/tm-core/src/common/utils/project-root-finder.ts`.
>   - **Constants**: Introduce path and marker constants in `common/constants/paths.ts` and export via `common/constants/index.ts` and `src/index.ts`.
>   - **Utils Export**: Re-export new utils from `common/utils/index.ts`.
> - **CLI**:
>   - **Root Auto-detection**: Add `apps/cli/src/utils/project-root.ts` and update commands (`autopilot/*`, `list`, `next`, `show`, `set-status`, `export`, `start`) to use `getProjectRoot()` instead of `process.cwd()`/manual merging.
>   - **DX**: In `scripts/dev.js`, load `.env` from detected project root and preserve original CWD.
>   - **Legacy Wrapper**: Simplify `src/utils/path-utils.js` to delegate root/normalization to `@tm/core` and improve relative path resolution using original CWD.
> - **Tests**:
>   - Add comprehensive unit tests for project-root detection with Vitest in `project-root-finder.spec.ts` and adjust integration tests to isolate in OS temp dirs; remove obsolete Jest unit suite for legacy finder.
> - **Config**:
>   - Tighten Jest coverage thresholds (notably for `src/utils/**` and `src/middleware/**`).
> - **Release**:
>   - Changeset: patch note "prioritize .taskmaster in parent directories over other project markers".
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 558adaa5ee2e80e3408e1e8952de5185d226c08c. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Review by Crunchyman-ralph — CHANGES_REQUESTED** [review]
please rebase with next.

Also, I think you'll notice we're moving some of our stuff to apps/cli and apps/mcp, maybe we can make sure the util improvements here are applied to the util findProjectRoot in apps/* ?

**Comment by Crunchyman-ralph** [comment]
@bjcoombs applied my own requested changes. if its all good for you, let me know and I'll merge this

**Comment by bjcoombs** [comment]
✅ **Rebased onto latest main**

All changes look good to me! I've rebased the PR on the latest main (synced with upstream) and all CI checks are passing:

- ✅ Format Check
- ✅ Typecheck  
- ✅ Build
- ✅ Test
- ✅ Cursor Bugbot

Ready to merge when you are. Thanks for applying your requested changes! 🙏

File: apps/cli/src/utils/project-root.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ *
+ * // Use explicit path if provided
+ * const root = getProjectRoot('/explicit/path');
+ * ```
+ */
+export function getProjectRoot(projectPath?: string): string {
+	// If explicitly provided, use it
+	if (projectPath) {
+		return projectPath;
+	}
```
Comment (Crunchyman-ralph) [inline apps/cli/src/utils/project-root.ts:39]: maybe need to do a path.resolve here ? for example for windows paths ?

**Comment by bjcoombs** [comment]
Hi @Crunchyman-ralph  👋

I've addressed all the inline code review feedback from the bots. Here's a summary of the fixes:

## Changes Made:

1. **scripts/dev.js** - Removed global `process.chdir()` to prevent breaking relative path logic
   - Now uses `dotenv.config({ path: join(projectRoot, '.env') })` instead

2. **project-root-finder.ts** - Fixed fallback behavior to return `process.cwd()` 
   - Ensures consistent behavior matching test expectations
   - Returns valid, existing directory path

3. **autopilot commands** - Moved `getProjectRoot()` calls into try blocks
   - Errors are now properly caught by error formatters
   - Fixed variable scoping for error handlers

4. **autopilot/index.ts** - Removed early evaluation of `getProjectRoot()`
   - No longer evaluates at module load time
   - Project root now resolved during command execution

5. **jest.config.js** - Fixed coverage threshold globs from `.js` to `.ts`
   - Matches ts-jest's actual coverage reporting

6. **project-root.ts** - Added `path.resolve()` for Windows path compatibility
   - Handles relative paths correctly across platforms

7. **Code formatting** - Applied proper formatting to all modified files

## CI Status:
- ✅ Build passing
- ✅ Format Check passing
- ✅ Typecheck passing
- ⏳ Tests running

I think everything's been sorted. Let me know if you need any additional changes!

File: jest.config.js [hunk 2]
```diff
@@ -43,12 +43,26 @@ export default {
...
+			branches: 70,
+			functions: 90,
+			lines: 90,
+			statements: 90
+		},
+		'./src/middleware/**/*.ts': {
+			branches: 70,
+			functions: 85,
+			lines: 85,
+			statements: 85
```
Comment (Crunchyman-ralph) [inline jest.config.js:65]: we don't have a src/middleware, remove this please

**Review by Crunchyman-ralph — APPROVED** [review]
lgtm, one final comment, but can merge anyway
